walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,85 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    ArrayList<TimeRange> timeRanges = new ArrayList<TimeRange>();
+    ArrayList<Integer> startTimes = new ArrayList<Integer>();
+    ArrayList<Integer> endTimes = new ArrayList<Integer>();
+
+    // If the duration is longer than the entire day, return an empty collection. 
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+        return timeRanges;
+    }
+
+    // If there are no events or attendees in the request, return the whole day. 
+    if (events.isEmpty() || request.getAttendees().isEmpty()) {
+        timeRanges.add(TimeRange.WHOLE_DAY);
+        return timeRanges;
+    }
+
+    // Loop through the list of events to add the start and end times for an event of the requested attendees. 
+    for (Event event : events) {
+        for (String eventPerson : event.getAttendees()) {
+            if (request.getAttendees().contains(eventPerson)) {
+                startTimes.add(event.getWhen().start());
+                endTimes.add(event.getWhen().end());
+            } 
+        }
+    }
+
+    // Sort the start and end times to simplify further processing. 
+    Collections.sort(startTimes);
+    Collections.sort(endTimes);
+
+    // If the first start time does not begin with 0, include a time starting from 0 to the first start time. 
+    if (startTimes.get(0) != 0) {
+        timeRanges.add(TimeRange.fromStartEnd(0, startTimes.get(0), false));
+    }
+
+    // If the last end time does not end with 1440, include a time ending from 1440 with the last end time. 
+    if (endTimes.get(endTimes.size()-1) != 1440) {
+        timeRanges.add(TimeRange.fromStartEnd(endTimes.get(endTimes.size()-1), 1440, false));
+    }
+
+    // Loop through the end times to combine various availability times with the start times.
+    for (int i = 0; i < endTimes.size()-1; i++) {
+        if (endTimes.get(i) < startTimes.get(i+1)) {
+            timeRanges.add(TimeRange.fromStartEnd(endTimes.get(i), startTimes.get(i+1), false));
+        }
+    }
+
+    // Loops through the time ranges and removes overlapping times. 
+    Collections.sort(timeRanges, TimeRange.ORDER_BY_START);
+    for (int i = 0; i < timeRanges.size()-1; i++) { 
+        TimeRange currentRange = timeRanges.get(i);
+        TimeRange nextRange = timeRanges.get(i+1);
+        if (nextRange.overlaps(currentRange)) {
+            if (currentRange.contains(nextRange)) {
+                timeRanges.remove(currentRange);
+            } else {
+                timeRanges.remove(nextRange);
+            }
+        }
```

Comment:
Can you explain what is happening here? 

Also, we are modifying the array list within the loop. When we remove a timerange from the array list, all indices after it would shift by 1, so your for loop variable i is going to skip some elements.

Ex. <a, b, c, d>
list.get(1) => "b"
list.remove("b")
list.get(2) => "d"

"c" is skipped here. Is this intentional?

Reply:
Skipping elements was not intentional. I wanted to loop through the time ranges and remove any overlapping ranges, but I was not aware that it affected the loop counts. I have updated the code to remove the element from a different ArrayList instead.